lib/BackgroundJob/MailNotificationSender.php

```diff
@@ -0,0 +1,192 @@
+use OCP\IConfig;
+use OCP\IGroupManager;
+use OCP\ILogger;
+use OCP\IRequest;
+use OCP\IURLGenerator;
+use OCP\Mail\IMailer;
+use OCP\Share\Exceptions\ShareNotFound;
+use OCP\Share\IManager;
+use OCP\Notification\IManager as NotificationManager;
+
+/**
+ * Class MailNotificationSender
+ *
+ * @package OCA\Notifications\BackgroundJob
+ */
+class MailNotificationSender extends Job {
+
+	/** @var IManager  */
+	private $shareManager;
+
+	/** @var \OC\Group\Manager|IGroupManager  */
+	private $groupManager;
+
+	/** @var Manager  */
+	private $userManager;
+
+	/** @var NotificationManager  */
+	private $notificationManager;
+
+	/** @var IConfig  */
+	private $config;
+
+	/** @var IMailer  */
+	private $mailer;
+
+	/** @var NotificationMailerAdapter  */
+	private $mailerAdapter;
+
+	/** @var IURLGenerator  */
+	private $urlGenerator;
+
+	/** @var IRequest  */
+	private $request;
+
+	/** @var ILogger */
+	private $logger;
+
+	public function __construct(IManager $shareManager = null,
+								IGroupManager $groupManager = null,
+								Manager $userManager = null,
+								NotificationManager $notificationManager = null,
+								NotificationMailerAdapter $mailerAdapter = null,
+								IMailer $mailer = null,
+								IConfig $config = null,
+								IURLGenerator $urlGenerator = null,
+								IRequest $request = null,
+								ILogger $logger = null) {
+		$this->shareManager = $shareManager ? $shareManager : \OC::$server->getShareManager();
+		$this->groupManager = $groupManager ? $groupManager : \OC::$server->getGroupManager();
+		$this->userManager = $userManager ? $userManager : \OC::$server->getUserManager();
+		$this->notificationManager = $notificationManager ? $notificationManager : \OC::$server->getNotificationManager();
+		$this->urlGenerator = $urlGenerator ? $urlGenerator : \OC::$server->getURLGenerator();
+		$this->request = $request ? $request : \OC::$server->getRequest();
+		$this->logger = $logger ? $logger : \OC::$server->getLogger();
+		$this->mailer = $mailer ? $mailer : \OC::$server->getMailer();
+		$this->config = $config ? $config : \OC::$server->getConfig();
+
+		$optionsStorage = new OptionsStorage($this->config);
+		$notificationMailer = new NotificationMailer($this->notificationManager, $this->mailer, $optionsStorage);
+
+		$this->mailerAdapter = $mailerAdapter ? $mailerAdapter :
+			new NotificationMailerAdapter($notificationMailer, $this->userManager, $this->logger, $this->urlGenerator);
+	}
+
+	/**
+	 * @param $notificationURL
+	 * @return string
+	 */
+	public function fixURL($notificationURL) {
+		$url = $this->request->getServerProtocol() . '://' . $this->request->getServerHost();
+		$partURL = \explode($url, $notificationURL)[1];
+		$webRoot = $this->getArgument()['webroot'];
+		if (\strpos($partURL, $webRoot) === false) {
+			$notificationURL = $url . $this->getArgument()['webroot'] . $partURL;
+		}
+		return $notificationURL;
+	}
+
+	/**
+	 * @param $shareFullId
+	 */
+	public function sendNotify($shareFullId) {
```

Comment:
unless there is a reason to call this method from the outside, keep the method private